src/interface_types/resource_handles.rs

```diff
@@ -0,0 +1,371 @@
+// Copyright 2020 Contributors to the Parsec project.
+// SPDX-License-Identifier: Apache-2.0
+use crate::{
+    constants::tss::{
+        TPM2_RH_ENDORSEMENT, TPM2_RH_LOCKOUT, TPM2_RH_NULL, TPM2_RH_OWNER, TPM2_RH_PLATFORM,
+        TPM2_RH_PLATFORM_NV,
+    },
+    handles::{AuthHandle, NvIndexHandle, TpmConstantsHandle},
+    tss2_esys::{
+        ESYS_TR, ESYS_TR_RH_ENDORSEMENT, ESYS_TR_RH_LOCKOUT, ESYS_TR_RH_NULL, ESYS_TR_RH_OWNER,
+        ESYS_TR_RH_PLATFORM, ESYS_TR_RH_PLATFORM_NV, TPM2_HANDLE, TPM2_RH, TPMI_RH_HIERARCHY,
+    },
+    Error, Result, WrapperErrorKind,
+};
+use std::convert::TryFrom;
+//////////////////////////////////////////////////////////////////////////////////
+/// Hierarchy
+///
+/// Enum describing the object hierarchies in a TPM 2.0.
+//////////////////////////////////////////////////////////////////////////////////
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum Hierarchy {
+    Owner,
+    Platform,
+    Endorsement,
+    Null,
+}
+
+impl Hierarchy {
+    /// Get the ESYS resource handle for the hierarchy.
+    pub fn esys_rh(self) -> TPMI_RH_HIERARCHY {
+        match self {
+            Hierarchy::Owner => ESYS_TR_RH_OWNER,
+            Hierarchy::Platform => ESYS_TR_RH_PLATFORM,
+            Hierarchy::Endorsement => ESYS_TR_RH_ENDORSEMENT,
+            Hierarchy::Null => ESYS_TR_RH_NULL,
+        }
+    }
+
+    /// Get the TPM resource handle for the hierarchy.
+    pub fn rh(self) -> TPM2_RH {
+        match self {
+            Hierarchy::Owner => TPM2_RH_OWNER,
+            Hierarchy::Platform => TPM2_RH_PLATFORM,
+            Hierarchy::Endorsement => TPM2_RH_ENDORSEMENT,
+            Hierarchy::Null => TPM2_RH_NULL,
+        }
+    }
+}
+
+impl TryFrom<TPM2_HANDLE> for Hierarchy {
+    type Error = Error;
+
+    fn try_from(handle: TPM2_HANDLE) -> Result<Self> {
+        match handle {
+            TPM2_RH_OWNER | ESYS_TR_RH_OWNER => Ok(Hierarchy::Owner),
+            TPM2_RH_PLATFORM | ESYS_TR_RH_PLATFORM => Ok(Hierarchy::Platform),
+            TPM2_RH_ENDORSEMENT | ESYS_TR_RH_ENDORSEMENT => Ok(Hierarchy::Endorsement),
+            TPM2_RH_NULL | ESYS_TR_RH_NULL => Ok(Hierarchy::Null),
+            _ => Err(Error::local_error(WrapperErrorKind::InconsistentParams)),
```

Comment:
This should probabily be `InvalidParams` (and similarly for the other `TryFrom`s)